Add Support for OAuth in athena-synapse Connector

Jithendar12 opened the pull request:
*Issue #, if available:*

*Description of changes:*
This PR introduces support for OAuth-based authentication in the Synapse connector using Microsoft Entra ID with the client_credentials grant type. The connector obtains Bearer tokens directly from Entra ID and manages them securely via AWS Secrets Manager.

Users must store the following credentials in AWS Secrets Manager:
{
  "client_id": "&lt;your-entra-id-app-client-id>",
  "client_secret": "&lt;your-entra-id-app-client-secret>",
  "tenant_id": "&lt;your-entra-id-app-tenant-id>"
}

The connector uses these credentials to fetch a new access token from Microsoft Entra ID's token endpoint. After fetching, it updates the secret with access_token, fetched_at and expires_in. Token refresh is handled automatically when the token is close to expiry, using these stored values. Please find attached reference and test documents.
[OAuth Integration in synapse.docx](https://github.com/user-attachments/files/21509956/OAuth.Integration.in.synapse.docx)
[synapse-oauth-testing.odt](https://github.com/user-attachments/files/21509960/synapse-oauth-testing.odt)

By submitting this pull request, I confirm that you can use, modify, copy, and redistribute this contribution, under the terms of your choice.

AbdulR3hman commented on athena-synapse/src/main/java/com/amazonaws/athena/connectors/synapse/SynapseCredentialsProvider.java:
```diff
@@ -0,0 +1,211 @@
+
+            return credentialMap;
+        }
+        catch (Exception e) {
```
Recommendation generated by Amazon CodeGuru Reviewer. Leave feedback on this recommendation by replying to the comment or by reacting to the comment using emoji.

It appears that your code handles a broad swath of exceptions in the catch block, potentially trapping dissimilar issues or problems that should not be dealt with at this point in the program.

[Learn more](https://cwe.mitre.org/data/definitions/396.html)

AbdulR3hman commented on athena-synapse/src/main/java/com/amazonaws/athena/connectors/synapse/SynapseCredentialsProvider.java:
```diff
@@ -0,0 +1,211 @@
+            }
+            return null;
+        }
+        catch (Exception e) {
```
Recommendation generated by Amazon CodeGuru Reviewer. Leave feedback on this recommendation by replying to the comment or by reacting to the comment using emoji.

It appears that your code handles a broad swath of exceptions in the catch block, potentially trapping dissimilar issues or problems that should not be dealt with at this point in the program.

[Learn more](https://cwe.mitre.org/data/definitions/396.html)

AbdulR3hman commented on athena-synapse/src/main/java/com/amazonaws/athena/connectors/synapse/SynapseCredentialsProvider.java:
```diff
@@ -0,0 +1,211 @@
+                !oauthConfig.get(TENANT_ID).isEmpty();
+    }
+
+    private String fetchAccessToken(Map<String, String> oauthConfig) throws Exception
```
Recommendation generated by Amazon CodeGuru Reviewer. Leave feedback on this recommendation by replying to the comment or by reacting to the comment using emoji.

Methods are throwing general exception, instead try and throw specific exception for proper and easy error handling.

Jithendar12 replied:
Replaced generic Exception with specific exceptions (IOException, InterruptedException) to improve error handling and clarity.

AbdulR3hman commented on athena-synapse/src/main/java/com/amazonaws/athena/connectors/synapse/SynapseCredentialsProvider.java:
```diff
@@ -0,0 +1,211 @@
+        return fetchAndStoreNewToken(oauthConfig);
+    }
+
+    private String fetchAndStoreNewToken(Map<String, String> oauthConfig) throws Exception
```
Recommendation generated by Amazon CodeGuru Reviewer. Leave feedback on this recommendation by replying to the comment or by reacting to the comment using emoji.

Methods are throwing general exception, instead try and throw specific exception for proper and easy error handling.

Jithendar12 replied:
Replaced generic Exception with specific exceptions (IOException, InterruptedException) to improve error handling and clarity.

AbdulR3hman commented on athena-synapse/src/main/java/com/amazonaws/athena/connectors/synapse/SynapseCredentialsProvider.java:
```diff
@@ -0,0 +1,211 @@
+    private static final String FETCHED_AT = "fetched_at";
+    private static final String EXPIRES_IN = "expires_in";
+
+    private static final String CLIENT_ID = "client_id";
```
Recommendation generated by Amazon CodeGuru Reviewer. Leave feedback on this recommendation by replying to the comment or by reacting to the comment using emoji.

Potential hardcoded credential detected. This code may contain sensitive data such as passwords or API keys embedded directly in the source. Hardcoded credentials can be extracted and misused, leading to unauthorized access to systems or data breaches. To remediate this, store secrets in environment variables or use a secrets management tool like AWS Secrets Manager, Azure Key Vault, or HashiCorp Vault. Avoid committing credentials to version control. For best practices, refer to - https://cwe.mitre.org/data/definitions/798.html

AbdulR3hman commented on athena-synapse/src/main/java/com/amazonaws/athena/connectors/synapse/SynapseCredentialsProvider.java:
```diff
@@ -0,0 +1,211 @@
+    private static final String TENANT_ID = "tenant_id";
+
+    private static final String USER = "user";
+    private static final String PASSWORD = "password";
```
Recommendation generated by Amazon CodeGuru Reviewer. Leave feedback on this recommendation by replying to the comment or by reacting to the comment using emoji.

Potential hardcoded credential detected. This code may contain sensitive data such as passwords or API keys embedded directly in the source. Hardcoded credentials can be extracted and misused, leading to unauthorized access to systems or data breaches. To remediate this, store secrets in environment variables or use a secrets management tool like AWS Secrets Manager, Azure Key Vault, or HashiCorp Vault. Avoid committing credentials to version control. For best practices, refer to - https://cwe.mitre.org/data/definitions/798.html

Jithendar12 replied:
These are just constant field names used as keys to access the configuration, not actual credentials. The actual sensitive data is stored securely in AWS Secrets Manager.

AbdulR3hman commented on athena-synapse/src/main/java/com/amazonaws/athena/connectors/synapse/SynapseCredentialsProvider.java:
```diff
@@ -0,0 +1,211 @@
+    private static final String EXPIRES_IN = "expires_in";
+
+    private static final String CLIENT_ID = "client_id";
+    private static final String CLIENT_SECRET = "client_secret";
```
Recommendation generated by Amazon CodeGuru Reviewer. Leave feedback on this recommendation by replying to the comment or by reacting to the comment using emoji.

Potential hardcoded credential detected. This code may contain sensitive data such as passwords or API keys embedded directly in the source. Hardcoded credentials can be extracted and misused, leading to unauthorized access to systems or data breaches. To remediate this, store secrets in environment variables or use a secrets management tool like AWS Secrets Manager, Azure Key Vault, or HashiCorp Vault. Avoid committing credentials to version control. For best practices, refer to - https://cwe.mitre.org/data/definitions/798.html

Jithendar12 replied:
These are just constant field names used as keys to access the configuration, not actual credentials. The actual sensitive data is stored securely in AWS Secrets Manager.

AbdulR3hman commented on athena-synapse/src/main/java/com/amazonaws/athena/connectors/synapse/SynapseCredentialsProvider.java:
```diff
@@ -0,0 +1,211 @@
+
+public class SynapseCredentialsProvider implements CredentialsProvider
+{
+    private static final String ACCESS_TOKEN = "access_token";
```
Recommendation generated by Amazon CodeGuru Reviewer. Leave feedback on this recommendation by replying to the comment or by reacting to the comment using emoji.

Potential hardcoded credential detected. This code may contain sensitive data such as passwords or API keys embedded directly in the source. Hardcoded credentials can be extracted and misused, leading to unauthorized access to systems or data breaches. To remediate this, store secrets in environment variables or use a secrets management tool like AWS Secrets Manager, Azure Key Vault, or HashiCorp Vault. Avoid committing credentials to version control. For best practices, refer to - https://cwe.mitre.org/data/definitions/798.html

Jithendar12 commented:
Addressed review feedback:

1. Regarding broad exception handling:
   - Updated to use specific exceptions with appropriate FederationSourceErrorCode values
   - Added proper error mapping for different scenarios (INVALID_RESPONSE_EXCEPTION, ENTITY_NOT_FOUND_EXCEPTION, etc.)
   - Removed generic catch blocks

2. Regarding methods throwing general exceptions:
   - Methods now throw specific exceptions (ResourceNotFoundException, JsonProcessingException, etc.)
   - Added proper error code mapping based on exception type

3. Regarding hardcoded credential constants:
   - These are just constant field names used as keys to access the configuration
   - No actual credentials are hardcoded, all sensitive data is stored securely in AWS Secrets Manager

4. Fixed failing test case:
   - Updated test to use proper ResourceNotFoundException
   - Updated assertions to match new error codes

Jithendar12 commented:
@AbdulR3hman Regarding the broad exception handling:
Updated to handle specific exceptions with appropriate error codes:
- JsonProcessingException → INVALID_RESPONSE_EXCEPTION
- ResourceNotFoundException → ENTITY_NOT_FOUND_EXCEPTION
- SecretsManagerException → INTERNAL_SERVICE_EXCEPTION
- IOException → OPERATION_TIMEOUT_EXCEPTION
- InterruptedException → OPERATION_TIMEOUT_EXCEPTION

Each exception type now maps to a specific error code for better error handling and debugging.

Jithendar12 commented on athena-synapse/src/main/java/com/amazonaws/athena/connectors/synapse/SynapseCredentialsProvider.java:
```diff
@@ -0,0 +1,312 @@
+                .timeout(Duration.ofSeconds(30))
+                .build();
+
+        HttpResponse<String> response = client.send(request, HttpResponse.BodyHandlers.ofString());
```
Hi @ejeffrli,
We’ve tested the connector within a VPC as well, and it is successfully able to retrieve the token and fetch the data.

ejeffrli commented on athena-synapse/src/main/java/com/amazonaws/athena/connectors/synapse/SynapseCredentialsProvider.java:
https://github.com/awslabs/aws-athena-query-federation/pull/2932/files#diff-58230124228b827596fd5cee17ea71175ded843859f3e26d9d9b2bbbc5110b99

Can this class use the OAuth class here? I can review above first if that is the case.

Jithendar12 replied:
Hi @ejeffrli, yes, please review [#2932](https://github.com/awslabs/aws-athena-query-federation/pull/2932) PR. Once that's approved and merged , I'll incorporate those changes into this PR and the [sap hana](https://github.com/awslabs/aws-athena-query-federation/pull/2894) PR. Thank you!

burhan94 commented:
This PR will change a lot after https://github.com/awslabs/aws-athena-query-federation/pull/2932 is merged, adding this comment as a reminder to myself and other reviewers

burhan94 commented on athena-synapse/src/main/java/com/amazonaws/athena/connectors/synapse/SynapseJdbcConnectionFactory.java:
```diff
@@ -60,25 +61,41 @@ public Connection getConnection(final CredentialsProvider credentialsProvider)
         try {
             final String derivedJdbcString;
-            if (null != credentialsProvider) {
+            final Properties connectionProps = new Properties();
```
Is there a reason why we still need the SynapseJdbcConnectionFactory? Can we not use the GenericConnectionFactory like we do in the other connectors now?

burhan94 commented on athena-synapse/src/main/java/com/amazonaws/athena/connectors/synapse/SynapseJdbcConnectionFactory.java:
```diff
@@ -60,25 +61,41 @@ public Connection getConnection(final CredentialsProvider credentialsProvider)
-                        "password=" + credentialsProvider.getCredentialMap().get(CredentialsConstants.PASSWORD)
-                    );
+                    // Check if this is OAuth credentials
+                    if (credentialsProvider.getCredential() instanceof OAuthAccessTokenCredentials) {
```
hmm, the whole point of the abstraction we did was so that we don't need to put a specific credential provider type here...

VenkatasivareddyTR replied:
Hi @burhan94, yes, code has been updated, please review it. Thank you!

VenkatasivareddyTR commented on athena-synapse/src/main/java/com/amazonaws/athena/connectors/synapse/SynapseCredentialsProvider.java:
```diff
@@ -0,0 +1,312 @@
+    // Constants for basic authentication fields
+    private static final String USER = "user";
+    private static final String PASSWORD = "password";
+    private static final String USERNAME = "username";
```
Hi @burhan94, yes, we have moved to common CredentialsConstants file. Thank you!

Jithendar12 commented on athena-synapse/src/main/java/com/amazonaws/athena/connectors/synapse/SynapseJdbcConnectionFactory.java:
```diff
@@ -60,25 +60,29 @@ public Connection getConnection(final CredentialsProvider credentialsProvider)
+
+                String connectionString = databaseConnectionConfig.getJdbcConnectionString();
+
+                if (connectionString.contains("authentication=ActiveDirectoryServicePrincipal")) {
```
Hi @burhan94 ,
Yes, we can move secret replacement into the CredentialsProvider interface. We’ve made this change; please review. The SynapseJdbcConnectionFactory class is no longer required, as we can use the GenericJdbcConnectionFactory class, so we’ve deleted it. We can apply the same approach to other connectors as well, but that can be handled as a separate task since it will require testing all connectors once this PR is merged. Thank you!
